api/src/controllers/validation.ts

```diff
@@ -1,51 +1,92 @@
 import { HttpError } from "./error";
-import { isInArray } from "../helpers";
+import { isInSet, not } from "../helpers";
+import { StringLiteral } from "../types";
 
-type StringLiteral<T> = T extends string
-  ? string extends T
-    ? never
-    : T
-  : never;
+type NonEmptyArray<T> = [T, ...T[]];
 
 type QueryValidatorConfig<Name, AllowedValue> = {
   name: StringLiteral<Name>;
-  defaultValue: Readonly<StringLiteral<AllowedValue>>;
   allowed: ReadonlyArray<StringLiteral<AllowedValue>>;
+  defaultValue?: Readonly<StringLiteral<AllowedValue>>;
+  singleValue?: boolean;
 };
-
-export const queryValidator =
-  <Name, AllowedValue>({
-    name,
-    defaultValue,
-    allowed,
-  }: QueryValidatorConfig<Name, AllowedValue>) =>
-  <Query extends { [key in typeof name]?: AllowedValue | string }>(
+export const queryValidator = <Name, AllowedValue>({
+  name,
+  allowed,
+  defaultValue,
+  singleValue,
+}: QueryValidatorConfig<Name, AllowedValue>) => {
+  const allowedSet = new Set(allowed);
+  return <Query extends { [key in typeof name]?: string }>(
     query: Query
-  ): AllowedValue => {
-    const value: AllowedValue | string | undefined = query[name];
-    if (typeof value === "undefined") {
-      return defaultValue;
-    } else if (isInArray(value, allowed)) {
-      return value;
-    } else {
+  ): NonEmptyArray<AllowedValue> | undefined => {
+    const providedValues = query[name]?.split(",");
+    if (providedValues === undefined) {
+      return defaultValue === undefined ? undefined : [defaultValue];
+    }
+
+    const validValues = providedValues.filter(isInSet(allowedSet));
+    if (
+      validValues.length !== providedValues.length ||
+      validValues.length === 0
+    ) {
```

Comment:
I see what you mean about missing Scala; this is the sort of thing where `.partition()` would come in really handy.